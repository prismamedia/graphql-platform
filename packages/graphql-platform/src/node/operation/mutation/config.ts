import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
