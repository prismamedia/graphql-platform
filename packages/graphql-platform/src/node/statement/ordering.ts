import { MGetter, MMethod } from '@prismamedia/memoize';
import * as graphql from 'graphql';
import * as R from 'remeda';
import type { Node } from '../../node.js';
import { DependencyGraph } from '../change/dependency.js';
import type { OrderByInputValue } from '../type.js';
import type { OrderingExpression } from './ordering/expression.js';

export * from './ordering/direction.js';
export * from './ordering/expression.js';

export class NodeOrdering {
  public readonly expressions: ReadonlyArray<OrderingExpression>;
  public readonly normalized: this | undefined;

  public constructor(
    public readonly node: Node,
    expressions: ReadonlyArray<OrderingExpression>,
  ) {
    this.expressions = R.uniqueWith(expressions, (a, b) => a.equals(b));

    this.normalized = this.expressions.length ? this : undefined;
  }

  public equals(ordering: unknown): boolean {
    return (
      ordering instanceof NodeOrdering &&
      ordering.node === this.node &&
      ordering.expressions.length === this.expressions.length &&
      ordering.expressions.every((expression, index) =>
        expression.equals(this.expressions[index]),
      )
    );
  }

  @MGetter
  public get dependencyGraph(): DependencyGraph {
    return new DependencyGraph(
      this.node,
      ...this.expressions.map(({ dependency }) => dependency),
    );
  }

  @MGetter
  public get ast(): graphql.ConstListValueNode {
    return {
      kind: graphql.Kind.LIST,
      values: this.expressions.map(({ ast }) => ast),
    };
  }

  @MMethod()
  public toString(): string {
    return graphql.print(this.ast);
  }

  @MGetter
  public get inputValue(): NonNullable<OrderByInputValue> {
    return this.expressions.map(({ inputValue }) => inputValue);
  }
}

export const areOrderingsEqual = (
  a: NodeOrdering | undefined,
  b: NodeOrdering | undefined,
): boolean =>
  a?.normalized && b?.normalized
    ? a.normalized.equals(b.normalized)
    : a?.normalized === b?.normalized;
