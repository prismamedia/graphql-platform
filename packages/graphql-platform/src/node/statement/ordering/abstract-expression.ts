import type * as graphql from 'graphql';
import type { Constructor } from 'type-fest';
import type { Dependency } from '../../change.js';
import type { NodeOrderingInputValue } from '../../type.js';
import type { OrderingDirection } from './direction.js';

export abstract class AbstractOrderingExpression {
  /**
   * Returns the AST for this expression
   */
  public abstract readonly ast: graphql.EnumValueNode;

  /**
   * Returns the input-value for this expression
   */
  public abstract readonly inputValue: NonNullable<NodeOrderingInputValue>;

  public abstract readonly dependency?: Dependency;

  public constructor(public readonly direction: OrderingDirection) {}

  public equals(expression: unknown): expression is this {
    return (
      expression instanceof (this.constructor as Constructor<this>) &&
      expression.direction === this.direction
    );
  }
}
