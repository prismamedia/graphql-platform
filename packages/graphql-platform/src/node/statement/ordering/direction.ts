export enum OrderingDirection {
  ASCENDING = 'ASCENDING',
  DESCENDING = 'DESCENDING',
}
