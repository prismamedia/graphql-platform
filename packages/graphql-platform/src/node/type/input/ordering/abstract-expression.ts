import * as utils from '@prismamedia/graphql-platform-utils';
import type { OperationContext } from '../../../operation.js';
import {
  OrderingDirection,
  type OrderingExpression,
} from '../../../statement/ordering.js';

export abstract class AbstractOrderingExpressionInput extends utils.EnumInputValue {
  public abstract readonly direction: OrderingDirection;

  public abstract sort(
    context?: OperationContext,
    path?: utils.Path,
  ): OrderingExpression;
}
