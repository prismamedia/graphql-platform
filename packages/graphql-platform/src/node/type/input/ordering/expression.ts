import { LeafOrderingInput } from './expression/leaf.js';
import { MultipleReverseEdgeCountOrderingInput } from './expression/reverse-edge-multiple-count.js';

export * from './expression/leaf.js';
export * from './expression/reverse-edge-multiple-count.js';

export type OrderingExpressionInput =
  | LeafOrderingInput
  | MultipleReverseEdgeCountOrderingInput;

export const isOrderingExpressionInput = (
  maybeOrderingExpressionInput: unknown,
): maybeOrderingExpressionInput is OrderingExpressionInput =>
  maybeOrderingExpressionInput instanceof LeafOrderingInput ||
  maybeOrderingExpressionInput instanceof MultipleReverseEdgeCountOrderingInput;
