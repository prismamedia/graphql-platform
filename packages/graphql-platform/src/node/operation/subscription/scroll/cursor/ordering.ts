import * as utils from '@prismamedia/graphql-platform-utils';
import { MGetter } from '@prismamedia/memoize';
import assert from 'node:assert';
import type { Node } from '../../../../../node.js';
import { Leaf } from '../../../../definition.js';
import { OrderingDirection } from '../../../../statement/ordering/direction.js';
import type { LeafOrderingInput } from '../../../../type/input/ordering/expression/leaf.js';

export type ScrollCursorOrderingInputValue = utils.Nillable<
  LeafOrderingInput['value']
>;

export class ScrollCursorOrderingInputType extends utils.EnumInputType<LeafOrderingInput> {
  public constructor(public readonly node: Node) {
    assert(
      node.uniqueConstraintSet
        .values()
        .some((uniqueConstraint) => uniqueConstraint.isScrollable()),
    );

    super({
      name: `${node}ScrollCursorOrderingInput`,
      description: `Order the "${node}"'s scroll`,
    });
  }

  @MGetter
  public override get enumValues(): ReadonlyArray<LeafOrderingInput> {
    return this.node.uniqueConstraintSet
      .values()
      .filter((uniqueConstraint) => uniqueConstraint.isScrollable())
      .flatMap((uniqueConstraint) =>
        uniqueConstraint.componentSet.values().flatMap((component) => {
          assert(component instanceof Leaf);

          return [
            component.getOrderingInput(OrderingDirection.ASCENDING),
            component.getOrderingInput(OrderingDirection.DESCENDING),
          ];
        }),
      )
      .toArray();
  }
}
