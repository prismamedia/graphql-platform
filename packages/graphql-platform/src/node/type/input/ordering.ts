import * as utils from '@prismamedia/graphql-platform-utils';
import { MGetter } from '@prismamedia/memoize';
import type { Node } from '../../../node.js';
import { Leaf } from '../../definition.js';
import type { OperationContext } from '../../operation/context.js';
import { NodeOrdering, OrderingDirection } from '../../statement/ordering.js';
import { OrderingExpressionInput } from './ordering/expression.js';

export * from './ordering/expression.js';
export * from './ordering/expression/leaf.js';
export * from './ordering/expression/reverse-edge-multiple-count.js';

export type NodeOrderingInputValue = utils.Nillable<
  OrderingExpressionInput['value']
>;

export type OrderByInputValue = utils.Nillable<
  NonNullable<NodeOrderingInputValue>[]
>;

export type NodeOrderingInputTypeOverride = {
  name?: string;
  description?: string;
};

export class NodeOrderingInputType extends utils.EnumInputType<OrderingExpressionInput> {
  public constructor(
    public readonly node: Node,
    override?: Partial<NodeOrderingInputTypeOverride>,
  ) {
    super({
      name: override?.name ?? `${node}OrderingInput`,
      description: override?.description ?? `Order the "${node}" nodes`,
    });
  }

  @MGetter
  public override get enumValues(): ReadonlyArray<OrderingExpressionInput> {
    return [
      ...this.node.componentSet
        .values()
        .flatMap<OrderingExpressionInput>((component) =>
          component instanceof Leaf && component.isSortable()
            ? [
                component.getOrderingInput(OrderingDirection.ASCENDING),
                component.getOrderingInput(OrderingDirection.DESCENDING),
              ]
            : [],
        ),
      ...this.node.multipleReverseEdgeSet
        .values()
        .flatMap<OrderingExpressionInput>((reverseEdge) => [
          reverseEdge.getOrderingInput(OrderingDirection.ASCENDING),
          reverseEdge.getOrderingInput(OrderingDirection.DESCENDING),
        ]),
    ];
  }

  public sort(
    value: OrderByInputValue,
    context?: OperationContext,
    path?: utils.Path,
  ): NodeOrdering {
    utils.assertNillableIterableObject<NodeOrderingInputValue>(value, path);

    return new NodeOrdering(
      this.node,
      value
        ? Array.from(value, (value, index) =>
            this.getEnumValue(value, utils.addPath(path, index)).sort(
              context,
              utils.addPath(path, value),
            ),
          )
        : [],
    );
  }
}
