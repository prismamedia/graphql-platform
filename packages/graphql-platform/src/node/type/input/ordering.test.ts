import { beforeAll, describe, expect, it } from '@jest/globals';
import { GraphQLEnumType, printType } from 'graphql';
import { GraphQLPlatform } from '../../../index.js';
import { MyGP, nodeNames, nodes } from '../../../__tests__/config.js';
import {
  OrderingDirection,
  OrderingExpression,
} from '../../statement/ordering.js';
import { NodeOrderingInputType, OrderByInputValue } from './ordering.js';

describe('NodeOrderingInputType', () => {
  let gp: MyGP;

  beforeAll(() => {
    gp = new GraphQLPlatform({ nodes });
  });

  describe('Definition', () => {
    it.each(nodeNames)('%s has an ordering input type', (nodeName) => {
      const node = gp.getNodeByName(nodeName);

      const orderingInputType = node.orderingInputType;
      expect(orderingInputType).toBeInstanceOf(NodeOrderingInputType);

      if (orderingInputType.isPublic()) {
        expect(orderingInputType.getGraphQLInputType()).toBeInstanceOf(
          GraphQLEnumType,
        );
        expect(
          printType(orderingInputType.getGraphQLInputType()),
        ).toMatchSnapshot(orderingInputType.name);
      } else {
        expect(() => orderingInputType.getGraphQLInputType()).toThrowError(
          `The "${nodeName}OrderingInput" input type is private`,
        );
      }
    });
  });

  describe('Runtime', () => {
    describe('Fails', () => {
      it.each([
        [
          'Article',
          ['_id_DESCS'],
          '/0 - Expects a value among "_id_ASC, _id_DESC, createdAt_ASC, createdAt_DESC, updatedAt_ASC, updatedAt_DESC, views_ASC, views_DESC, score_ASC, score_DESC, tagCount_ASC, tagCount_DESC", got: \'_id_DESCS\'',
        ],
      ])(
        'throws an Error on %sOrderingInput.sort(%p)',
        (nodeName, input, error) => {
          const node = gp.getNodeByName(nodeName);
          const orderingInputType = node.orderingInputType;

          expect(() => orderingInputType.sort(input)).toThrowError(error);
        },
      );
    });

    describe('Works', () => {
      it.each<
        [
          string,
          OrderByInputValue,
          ReadonlyArray<OrderingExpression['ast']> | undefined,
        ]
      >([
        ['Article', undefined, undefined],
        ['Article', null, undefined],
        ['Article', [], undefined],
        [
          'Article',
          ['createdAt_ASC'],
          [
            {
              kind: 'LeafOrdering',
              leaf: 'createdAt',
              direction: OrderingDirection.ASCENDING,
            },
          ],
        ],
        [
          'Article',
          ['tagCount_DESC', 'updatedAt_ASC'],
          [
            {
              kind: 'MultipleReverseEdgeCountOrdering',
              reverseEdge: 'tags',
              direction: OrderingDirection.DESCENDING,
            },
            {
              kind: 'LeafOrdering',
              leaf: 'updatedAt',
              direction: OrderingDirection.ASCENDING,
            },
          ],
        ],
      ])('%sOrderingInput.sort(%p)', (nodeName, input, expressions) => {
        const node = gp.getNodeByName(nodeName);
        const orderingInputType = node.orderingInputType;

        expect(
          orderingInputType.sort(input).normalized?.ast.expressions,
        ).toEqual(expressions);
      });
    });
  });
});
