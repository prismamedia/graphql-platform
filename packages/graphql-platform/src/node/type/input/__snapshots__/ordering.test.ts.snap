// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeOrderingInputType Definition Article has an ordering input type: ArticleOrderingInput 1`] = `
""""Order the "Article" nodes"""
enum ArticleOrderingInput {
  """from the lowest "createdAt" to the highest"""
  createdAt_ASC

  """from the highest "createdAt" to the lowest"""
  createdAt_DESC

  """from the lowest "updatedAt" to the highest"""
  updatedAt_ASC

  """from the highest "updatedAt" to the lowest"""
  updatedAt_DESC

  """from the lowest "views" to the highest"""
  views_ASC

  """from the highest "views" to the lowest"""
  views_DESC

  """from the lowest "score" to the highest"""
  score_ASC

  """from the highest "score" to the lowest"""
  score_DESC

  """from the lowest number of "tags" to the highest"""
  tagCount_ASC

  """from the highest number of "tags" to the lowest"""
  tagCount_DESC
}"
`;

exports[`NodeOrderingInputType Definition ArticleTag has an ordering input type: ArticleTagOrderingInput 1`] = `
""""Order the "ArticleTag" nodes"""
enum ArticleTagOrderingInput {
  """from the lowest "order" to the highest"""
  order_ASC

  """from the highest "order" to the lowest"""
  order_DESC

  """from the lowest number of "moderations" to the highest"""
  moderationCount_ASC

  """from the highest number of "moderations" to the lowest"""
  moderationCount_DESC
}"
`;

exports[`NodeOrderingInputType Definition Category has an ordering input type: CategoryOrderingInput 1`] = `
""""Order the "Category" nodes"""
enum CategoryOrderingInput {
  """from the lowest "order" to the highest"""
  order_ASC

  """from the highest "order" to the lowest"""
  order_DESC

  """from the lowest number of "children" to the highest"""
  childCount_ASC

  """from the highest number of "children" to the lowest"""
  childCount_DESC

  """from the lowest number of "articles" to the highest"""
  articleCount_ASC

  """from the highest number of "articles" to the lowest"""
  articleCount_DESC
}"
`;

exports[`NodeOrderingInputType Definition Tag has an ordering input type: TagOrderingInput 1`] = `
""""Order the "Tag" nodes"""
enum TagOrderingInput {
  """from the lowest "createdAt" to the highest"""
  createdAt_ASC

  """from the highest "createdAt" to the lowest"""
  createdAt_DESC

  """from the lowest "updatedAt" to the highest"""
  updatedAt_ASC

  """from the highest "updatedAt" to the lowest"""
  updatedAt_DESC

  """from the lowest number of "articles" to the highest"""
  articleCount_ASC

  """from the highest number of "articles" to the lowest"""
  articleCount_DESC
}"
`;

exports[`NodeOrderingInputType Definition User has an ordering input type: UserOrderingInput 1`] = `
""""Order the "User" nodes"""
enum UserOrderingInput {
  """from the lowest number of "createdArticles" to the highest"""
  createdArticleCount_ASC

  """from the highest number of "createdArticles" to the lowest"""
  createdArticleCount_DESC

  """from the lowest number of "updatedArticles" to the highest"""
  updatedArticleCount_ASC

  """from the highest number of "updatedArticles" to the lowest"""
  updatedArticleCount_DESC

  """from the lowest number of "articleTagModerations" to the highest"""
  articleTagModerationCount_ASC

  """from the highest number of "articleTagModerations" to the lowest"""
  articleTagModerationCount_DESC
}"
`;

exports[`NodeOrderingInputType Definition UserProfile has an ordering input type: UserProfileOrderingInput 1`] = `
""""Order the "UserProfile" nodes"""
enum UserProfileOrderingInput {
  """from the lowest "birthday" to the highest"""
  birthday_ASC

  """from the highest "birthday" to the lowest"""
  birthday_DESC
}"
`;
